adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -735,41 +946,210 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+          preservesPartitioning = true)
+      }
+
+      referenceRegionKeyedGenomicRDD.mapPartitions(iter => {
+        iter.flatMap(f => {
+          val intervals = partitionMapIntervals.get(f._1, requireOverlap = false)
+          intervals.map(g => ((f._1, g._2), f._2))
+        })
+      }, preservesPartitioning = true)
+        .repartitionAndSortWithinPartitions(
+          new ReferenceRegionRangePartitioner(numPartitions))
+        // return to an RDD[(ReferenceRegion, T)], removing the partition ID
+        .map(f => (f._1._1, f._2))
+    }
+    finalPartitionedRDD.cache()
+
+    // we cannot use the bounds of the left RDD because
+    // they may not necessarily be the bounds of the right RDD
+    val newPartitionMap = finalPartitionedRDD.mapPartitions(iter => {
+      getRegionBoundsFromPartition(iter)
+    }, preservesPartitioning = true)
+      // we want to try really hard to prevent None's in the partition map
+      // so we use the left RDD's partition bounds on a partition if there
+      // is no right data there
+      .zipWithIndex.map(f => {
+        if (f._1.isEmpty) {
+          // use the bounds from the left partition map
+          destinationPartitionMap(f._2.toInt)
+        } else {
+          f._1
+        }
+      }).collect
+
+    replaceRdd(finalPartitionedRDD.values, Some(newPartitionMap))
   }
+
+  /**
+   * Gets the partition bounds from a ReferenceRegion keyed Iterator
+   *
+   * @param iter The data on a given partition. ReferenceRegion keyed
+   * @return The bounds of the ReferenceRegions on that partition, in an Iterator
+   */
+  private def getRegionBoundsFromPartition(iter: Iterator[(ReferenceRegion, T)]): Iterator[Option[(ReferenceRegion, ReferenceRegion)]] = {
+    if (iter.isEmpty) {
+      // This means that there is no data on the partition, so we have no bounds
+      Iterator(None)
+    } else {
+      val firstRegion = iter.next
+      val lastRegion =
+        if (iter.hasNext) {
+          // we have to make sure we get the full bounds of this partition, this
+          // includes any extremely long regions. we include the firstRegion for
+          // the case that the first region is extremely long
+          (iter ++ Iterator(firstRegion)).maxBy(f => (f._1.referenceName, f._1.end, f._1.start))
+          // only one record on this partition, so this is the extent of the bounds
+        } else {
+          firstRegion
+        }
+      Iterator(Some((firstRegion._1, lastRegion._1)))
+    }
+  }
+
+  /**
+   * Sort and repartition the data. We use Sparks sortByKey on the
+   * ReferenceRegions of a given record. This gets us data that is
+   * partitioned to be roughly uniform in size across all nodes.
+   *
+   * @param partitions The number of partitions to split the data into
+   * @return A U that contains the sorted and partitioned RDD
+   */
+  def repartitionAndSort(partitions: Int = rdd.partitions.length)(implicit c: ClassTag[T]): U = {
```

Comment:
Important point of discussion: should a sort (and thus a join) still be allowed if the `GenomicRDD` has multiple `ReferenceRegions` for a record? Not allowing this would allow us to completely remove the `extractReferenceRegionsForPartition` method altogether. If we do allow it, `sort()` will have to be changed to allow for that. I have no opinion either way, nor do I know the right answer. @fnothaft @heuermh I would appreciate your inputs.